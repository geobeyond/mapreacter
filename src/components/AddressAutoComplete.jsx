import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types';
import keycode from 'keycode';
import Downshift from 'downshift';
import { withStyles } from 'material-ui/styles';
import TextField from 'material-ui/TextField';
import Paper from 'material-ui/Paper';
import { MenuItem } from 'material-ui/Menu';
import Chip from 'material-ui/Chip';
import Typography from 'material-ui/Typography';
import * as mapActions from '@boundlessgeo/sdk/actions/map';
import { mylocalizedstrings } from '../services/localizedstring';


var axios = require('axios');

const styles = theme => ({
  root: {
    margin: '15px',
    width: '400px',
    color: 'currentColor'
  },
  container: {
    flexGrow: 1,
    position: 'relative',
  },
  paper: {
    position: 'absolute',
    zIndex: 1,
    marginTop: theme.spacing.unit,
    left: 0,
    right: 0,
  },
  chip: {
    height: '25px',
    margin: '2px',
  },
  inputRoot: {
    flexWrap: 'wrap',
    color: 'currentColor'
  },
});



function renderInput(inputProps) {
  const { InputProps, classes, ref, ...other } = inputProps;
  console.log("AddressAutoComplete.renderInput()");
  return (
    <TextField
      InputProps={{
        inputRef: ref,
        classes: {
          root: classes.inputRoot,
        },
        ...InputProps,
      }}
      {...other}
    />
  );
}

function renderSuggestion({ suggestion, index, itemProps, highlightedIndex, selectedItem }) {
  console.log("AddressAutoComplete.renderSuggestion()", JSON.stringify(suggestion));
  const isHighlighted = highlightedIndex === index;
  const isSelected = (selectedItem || '').indexOf(suggestion.label) > -1;

  return (
    <MenuItem
      {...itemProps}
      key={suggestion.place_id}
      selected={isHighlighted}
      component="div"
      style={{
        fontWeight: isSelected ? 500 : 400,
      }}
    >
      <Typography variant="subheading">
        {suggestion.label}
      </Typography>
    </MenuItem>
  );
}
renderSuggestion.propTypes = {
  highlightedIndex: PropTypes.number,
  index: PropTypes.number,
  itemProps: PropTypes.object,
  selectedItem: PropTypes.string,
  suggestion: PropTypes.shape({ label: PropTypes.string }).isRequired,
};



class AddressAutoComplete extends React.Component {


  constructor(props) {
    super(props);
    console.log("AddressAutoComplete()");
    this.state = {
      inputValue: '',
      selectedItem: [],
      suggestions: []
    };
  }

  handleKeyDown = event => {
    console.log("AddressAutoComplete.handleKeyDown()");
    const { inputValue, selectedItem } = this.state;
    if (selectedItem.length && !inputValue.length && keycode(event) === 'backspace') {
      this.setState({
        selectedItem: selectedItem.slice(0, selectedItem.length - 1),
      });
    }
  };

  handleInputChange = event => {
    console.log("AddressAutoComplete.handleInputChange()", event.target.value);
    this.setState({ inputValue: event.target.value });

    const url = "https://nominatim.openstreetmap.org/search?format=json&polygon_geojson=1&q=" + event.target.value;
    console.log("GET", url);
    axios.get(url)
      .then((response) => {
        console.log("response:", JSON.stringify(response.data));      
        let _suggestion =  response.data.map( _record => {
          _record.label = _record.display_name;          
          return _record;
        })
        this.setState({ suggestions: _suggestion });
      })
      .catch((error) => {
        console.error(error);
      });
  };

  handleChange = item => {
    console.log("AddressAutoComplete.handleChange() item=", item);

    this.setState({
      inputValue: '',
      selectedItem: [item],
    });

    let selectedRecord = this.getSuggestions(item)[0];
    console.log("AddressAutoComplete.handleChange()", JSON.stringify(selectedRecord));
    this.props.removeFeatures("indirizzi");
    this.props.addFeatures("indirizzi", selectedRecord.geojson);
  };

  handleDelete = item => () => {
    const selectedItem = [...this.state.selectedItem];
    selectedItem.splice(selectedItem.indexOf(item), 1);
    this.setState({ selectedItem });
    console.log("RegProvAutocomplete.handleDelete()", item);
    this.props.removeFeatures("indirizzi");
  };

  getSuggestions(inputValue) {
    console.log("AddressAutoComplete.getSuggestions()", inputValue);
    let count = 0;
    return this.state.suggestions.filter(suggestion => {
      return (count++ < 12)
    });
  }

  render() {
    console.log("AddressAutoComplete.render()");
    const { classes } = this.props;
    const { inputValue, selectedItem } = this.state;

    return (
      <div className={classes.root}>
        <Downshift inputValue={inputValue} onChange={this.handleChange} selectedItem={selectedItem} classes={classes}>
          {({
            getInputProps,
            getItemProps,
            isOpen,
            inputValue: inputValue2,
            selectedItem: selectedItem2,
            highlightedIndex,
          }) => (
              <div className={classes.container}>
                {renderInput({
                  fullWidth: true,
                  classes,
                  InputProps: getInputProps({
                    startAdornment: selectedItem.map(item => (
                      <Chip
                        key={item}
                        tabIndex={-1}
                        label={item}
                        className={classes.chip}
                        onDelete={this.handleDelete(item)}
                      />
                    )),
                    onChange: this.handleInputChange,
                    onKeyDown: this.handleKeyDown,
                    placeholder: `${mylocalizedstrings.addressautocompletelabel}`,
                    id: 'integration-downshift-multiple',
                  }),
                })}
                {isOpen ? (
                  <Paper className={classes.paper} square>
                    {this.getSuggestions(inputValue2).map((suggestion, index) =>
                      renderSuggestion({
                        suggestion,
                        index,
                        itemProps: getItemProps({ item: suggestion.label }),
                        highlightedIndex,
                        selectedItem: selectedItem2,
                      }),
                    )}
                  </Paper>
                ) : null}
              </div>
            )}
        </Downshift>
      </div>
    );
  }
}

AddressAutoComplete.propTypes = {
  classes: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => {
  //console.log("AddressAutoComplete.mapStateToProps()");
  return {
    map: state.map,
    local: state.local,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addFeatures: (sourceName, features) => {
      dispatch(mapActions.addFeatures(sourceName, features));
    },
    removeFeatures: (sourceName, filter) => {
      dispatch(mapActions.removeFeatures(sourceName, filter));
    },
  };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(AddressAutoComplete)));