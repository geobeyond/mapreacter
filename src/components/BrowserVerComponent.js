import React, { Component } from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogTitle';
// eslint-disable-next-line
import DialogActions from '@material-ui/core/DialogActions';
// eslint-disable-next-line
import Button from '@material-ui/core/Button';
// eslint-disable-next-line
import { mylocalizedstrings } from '../services/localizedstring';

const theBrowserList = [
    {
        name: "Edge",
        minver: Number(16.0),
    },
    {
        name: "Chrome",
        minver: Number(57.0),
    },
    {
        name: "Safari",
        minver: Number(602.0),  // questa è la "webkit version" e corrisponde a Safari 10.0 
    },
    {
        name: "Opera",
        minver: Number(44.0),
    },
    {
        name: "Firefox",
        minver: Number(52.0),
    }
];

export var theBrowserItem = { name: "" };
export var theBrowserVersion = "";


class BrowserVerComponent extends Component {

    state = {
        sharedialog: false, 
        theMessage: ""
    };

    handleCloseMenu = () => {
        console.log("TocComponent.handleCloseMenu()");
        this.setState({ sharedialog: false });
    };

    componentDidMount() {

        //console.log("browser,  Browser CodeName: ", navigator.appCodeName);
        //console.log("browser,  Browser Name: ", navigator.appName);
        //console.log("browser,  Browser Version: ", navigator.appVersion);
        console.log("BrowserVerComponent.componentDidMount() Cookies Enabled: ", navigator.cookieEnabled);
        console.log("BrowserVerComponent.componentDidMount() Browser Language: ", navigator.language);
        console.log("BrowserVerComponent.componentDidMount() Browser Online: ", navigator.onLine);
        console.log("BrowserVerComponent.componentDidMount() Platform: ", navigator.platform);
        console.log("BrowserVerComponent.componentDidMount() User-agent header: ", navigator.userAgent);

        if (navigator.userAgent.indexOf("rv:11") > -1) {
            theBrowserItem = {
                name: "IE",
            }
            theBrowserVersion = Number(11);
            let theMessage = "il browser corrente (" + theBrowserItem.name + " " + theBrowserVersion + ") non è compatibile con l’applicazione ... " +
                "per una corretta visualizzazione è necessario utilizzare un browser di ultima generazione";
            this.setState({ sharedialog: true, theMessage });

        } else if (navigator.userAgent.indexOf("MSIE") > -1) {
            try {
                theBrowserItem = {
                    name: "MSIE",
                }
                let userAgentRecord = navigator.userAgent.split(";").find(function (element) {
                    return element.indexOf(theBrowserItem.name) > -1;
                });
                console.log("BrowserVerComponent.componentDidMount() userAgentRecord ->", userAgentRecord);
                let fullversion = userAgentRecord.split(" ")[2];
                console.log("BrowserVerComponent.componentDidMount() fullversion ->", fullversion);
                theBrowserVersion = Number(fullversion);
            } catch (error) {
            }
            let theMessage = "il browser corrente (" + theBrowserItem.name + " " + theBrowserVersion + ") non è compatibile con l’applicazione ... " +
                "per una corretta visualizzazione è necessario utilizzare un browser di ultima generazione";
            this.setState({ sharedialog: true, theMessage });

        } else {
            try {
                theBrowserItem = theBrowserList.find(function (element) {
                    return navigator.userAgent.indexOf(element.name) > -1;
                });

                let userAgentRecord = navigator.userAgent.split(" ").find(function (element) {
                    return element.indexOf(theBrowserItem.name) > -1;
                });
                let fullversion = userAgentRecord.split("/")[1];
                while (!Number(fullversion)) {
                    fullversion = fullversion.substr(0, fullversion.lastIndexOf("."));
                }
                theBrowserVersion = Number(fullversion);
                if (theBrowserVersion < theBrowserItem.minver) {
                    let theMessage = "il browser corrente (" + theBrowserItem.name + " " + theBrowserVersion + ") non è compatibile con l’applicazione, " +
                        "per una corretta visualizzazione è necessario eseguire l’aggiornamento alla versione " + theBrowserItem.minver + " o successive";
                    this.setState({ sharedialog: true, theMessage });
                }
            } catch (error) {
            }
        }
        console.log("BrowserVerComponent.componentDidMount() detected:", JSON.stringify(theBrowserItem), theBrowserVersion, this.state.sharedialog, this.state.theMessage);
    }

    render() {
        console.log("BrowserVerComponent.render()");
        return (
            <Dialog open={this.state.sharedialog} >
                <DialogContent style={{
                    //padding: '10px' 
                }}>
                    <h2>Attenzione</h2>
                    <p>{this.state.theMessage}</p>
                </DialogContent>
                <DialogActions>
                    <Button onClick={() => { this.handleCloseMenu(); }}>
                        {mylocalizedstrings.close}
                    </Button>
                </DialogActions>
            </Dialog>
        );
    }
}

export default BrowserVerComponent;

