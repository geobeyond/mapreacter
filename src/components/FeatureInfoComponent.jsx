import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import Dialog, { DialogContent, DialogActions } from 'material-ui/Dialog';
import Tabs, { Tab } from 'material-ui/Tabs';
import Table, { TableBody, TableRow, TableCell } from 'material-ui/Table';
import Button from 'material-ui/Button';
import * as actions from '../actions/map';
import { mylocalizedstrings } from '../services/localizedstring';


class FeatureInfoComponent extends Component {

    state = {
        value: 0,
    };

    handleChange = (event, value) => {
        this.setState({ value });
    };

    handleClose = () => {
        this.props.changeFeatureInfoComponent({ open: false });
    };

    renderBody(value) {
        console.log("FeatureInfoComponent.renderBody() value=", value);
        let rows = [];
        try {
            this.props.featureInfoComponent.items[value].features.forEach((feature, index) => {
                const _keys = Object.keys(feature.properties);
                _keys.forEach(_key => {
                    rows.push(
                        <TableRow key={_key}>
                            <TableCell>{_key}</TableCell>
                            <TableCell>{feature.properties[_key]}</TableCell>
                        </TableRow>
                    );
                });
            });
        } catch (error) {
            console.error(error);
        }
        return rows;
    }

    render() {
        console.log("FeatureInfoComponent.render()", this.props.featureInfoComponent.open);
        const { value } = this.state;
        if (!this.props.featureInfoComponent.open) { return null }
        return (
            <Dialog open={this.props.featureInfoComponent.open} >
                <DialogContent>
                    <Tabs
                        //indicatorColor="primary"
                        //textColor="primary"
                        scrollable
                        scrollButtons="on"
                        value={value}
                        onChange={this.handleChange}
                    >
                        {
                            this.props.featureInfoComponent.items.map(item => (
                                <Tab key={item.layer} label={item.layer} />
                            ))
                        }
                    </Tabs>
                    <Table>
                        <TableBody>{this.renderBody(value)}</TableBody>
                    </Table>
                </DialogContent>
                <DialogActions>
                    <Button onClick={this.handleClose}>{mylocalizedstrings.close}</Button>
                </DialogActions>
            </Dialog>
        )
    }

    componentDidCatch(error, info) {
        console.error(error, info);
    }
}

const mapStateToProps = (state) => {
    return {
        featureInfoComponent: state.local.featureInfoComponent,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        changeFeatureInfoComponent: (params) => {
            dispatch(actions.changeFeatureInfoComponent(params));
        },
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(FeatureInfoComponent));