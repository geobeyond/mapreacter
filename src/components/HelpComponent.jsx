import React, { Component } from 'react';
import { mylocalizedstrings } from '../services/localizedstring';
import contentIt from '../help/help_it.html';
import contentEn from '../help/help_en.html';

class HelpComponent extends Component {
    render() {
        if (mylocalizedstrings.getLanguage() === 'it') {
            return (
                <div dangerouslySetInnerHTML={{ __html: contentIt }} />
            );
        } else {
            return (
                <div dangerouslySetInnerHTML={{ __html: contentEn }} />
            );
        }
    }
}

export default HelpComponent