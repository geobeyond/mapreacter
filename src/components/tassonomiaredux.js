import { createStore } from 'redux';

const initialState = {
    phylumarr: [],
    famigliaarr: [],
    nomescientificoarr: [],
};

const NEWDATA = "NEWDATA";

export const newDataAction = (_data) => {
  return {
    type: NEWDATA,
    payload: { _data: _data }
  };
}

const tassonomiareducer = (state = initialState, action) => {
    switch (action.type) {
        case NEWDATA:
            state = Object.assign({}, state, {
                phylumarr: action.payload._data.phylum,
                famigliaarr: action.payload._data.famiglia,
                nomescientificoarr: action.payload._data.nome_scientifico,
            });
            break;
    }
    return state;
}

export const tassonomiastore = createStore(tassonomiareducer, initialState);

tassonomiastore.subscribe(() => {
    console.log(JSON.stringify(tassonomiastore.getState()))
});

