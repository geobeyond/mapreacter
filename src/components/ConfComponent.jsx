import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'

import * as mapActions from '@boundlessgeo/sdk/actions/map';
import * as printActions from '@boundlessgeo/sdk/actions/print';
import * as drawingActions from '@boundlessgeo/sdk/actions/drawing';
import { INTERACTIONS } from '@boundlessgeo/sdk/constants';

import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import IconButton from '@material-ui/core/IconButton';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogTitle from '@material-ui/core/DialogTitle';
import Button from '@material-ui/core/Button';

import * as configActions from '../actions/map';
import { mylocalizedstrings } from '../services/localizedstring';
import { downloadFile } from '../services/download';
import { theBrowserItem, theBrowserVersion } from './BrowserVerComponent';


class ConfComponent extends Component {

    state = {
        sharedialog: false,
        anchorEl: null
    };

    handleOpenMenu = event => {
        this.setState({ anchorEl: event.currentTarget });
    };

    handleCloseMenu = () => {
        this.setState({ anchorEl: null });
    };

    componentDidMount() {
        console.log("ConfComponent.componentDidMount() this.props=", JSON.stringify(this.props));
    }

    getActiveLayers() {
        let _array = [];
        this.props.layers.forEach((rec) => {
            if (rec.flag_filter) {
                if (rec['layout']) {
                    if (rec.layout.visibility === 'visible') {
                        _array.push(rec.id);
                    }
                }
            }
        });
        return _array;
    }

    render() {
        console.log("ConfComponent.render()");
        const { anchorEl } = this.state;
        return (
            <div>
                <IconButton onClick={this.handleOpenMenu}>
                    <i className="material-icons">more_vert</i>
                </IconButton>

                <Dialog
                    open={this.state.sharedialog}
                    onClose={() => { this.setState({ sharedialog: false }); }}
                >
                    <DialogTitle>{mylocalizedstrings.sharetitle}</DialogTitle>
                    {/*
                    <DialogContent>
                        <DialogContentText id="alert-dialog-description">
                            {mylocalizedstrings.sharetitle}
                        </DialogContentText>
                    </DialogContent> */}
                    <DialogActions>
                        <Button onClick={() => { this.setState({ sharedialog: false }); }}>
                            {mylocalizedstrings.close}
                        </Button>
                    </DialogActions>
                </Dialog>

                <Menu
                    id="menu-conf"
                    anchorEl={anchorEl}
                    anchorOrigin={{
                        vertical: 'top',
                        horizontal: 'right',
                    }}
                    transformOrigin={{
                        vertical: 'top',
                        horizontal: 'right',
                    }}
                    open={Boolean(anchorEl)}
                    onClose={this.handleCloseMenu}
                >
                    <MenuItem onClick={(event) => {
                        let url = this.props.local.mapConfig.helpUrl + mylocalizedstrings.getLanguage() + this.props.local.mapConfig.helpDoc;
                        let win = window.open(url, '_blank');
                        win.focus();
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">help</i>
                    </MenuItem>

                    <MenuItem onClick={(event) => {
                        var _txt = document.createElement('textarea');
                        _txt.value = window.location.href; //chrome
                        _txt.textContent = window.location.href; //firefox
                        document.body.appendChild(_txt);
                        _txt.select();
                        document.execCommand('copy');
                        console.log(_txt);
                        document.body.removeChild(_txt);
                        this.setState({ sharedialog: true });
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">share</i>
                    </MenuItem>


                    <MenuItem
                        disabled={navigator.userAgent.indexOf("Edge")>-1}
                        onClick={(event) => {
                            this.props.exportMapImage();
                            this.handleCloseMenu();
                        }}>
                        <i className="material-icons">file_download</i><span style={{ padding: '10px' }}>PNG</span>
                    </MenuItem>

                    <MenuItem onClick={(event) => {
                        downloadFile(this.props.local.mapConfig.downloadCSVUrl, this.getActiveLayers(), '.csv', this.props.local.regProvComponent['filter']);
                        this.handleCloseMenu();
                    }}>
                        <i className="material-icons">file_download</i><span style={{ padding: '10px' }}>CSV</span>
                    </MenuItem>

                    <MenuItem
                        disabled={navigator.userAgent.indexOf("Edge")>-1 && this.props.local.regProvComponent['filter'] && this.props.local.regProvComponent.filter.length>2000}
                        onClick={(event) => {
                            downloadFile(this.props.local.mapConfig.downloadShapefileUrl, this.getActiveLayers(), '.zip', this.props.local.regProvComponent['filter']);
                            this.handleCloseMenu();
                        }}>
                        <i className="material-icons">file_download</i><span style={{ padding: '10px' }}>Shapefile</span>
                    </MenuItem>

                    { /*
                    <MenuItem onClick={(event) => {
                        downloadFile(this.props.local.mapConfig.downloadPdfUrl, this.getActiveLayers(), '.pdf', this.props.local.regProvComponent['filter']);
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">file_download</i><span style={{ padding: '10px' }}>PDF</span>
                    </MenuItem>
                    */ }

                    <MenuItem onClick={(event) => {
                        this.props.startMeasure(INTERACTIONS.measure_line);
                        this.props.changeMeasureComponent({ open: true });
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">photo_size_select_small</i><span style={{ padding: '10px' }}>{mylocalizedstrings.line}</span>
                    </MenuItem>

                    <MenuItem onClick={(event) => {
                        this.props.startMeasure(INTERACTIONS.measure_polygon);
                        this.props.changeMeasureComponent({ open: true });
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">photo_size_select_small</i><span style={{ padding: '10px' }}>{mylocalizedstrings.polygon}</span>
                    </MenuItem>

                    <MenuItem onClick={(event) => {
                        let zoom = this.props.local.mapConfig.map.zoom || 2;
                        this.props.setView(this.props.local.mapConfig.map.center, zoom);
                        this.handleCloseMenu();
                    }} >
                        <i className="material-icons">fullscreen</i>
                    </MenuItem>

                    <MenuItem disabled style={{ fontStyle: 'italic', paddingTop: 0, paddingBottom: 0 }}>
                        {theBrowserItem.name} {theBrowserVersion}
                    </MenuItem>
                </Menu>

            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        local: state.local,
        layers: state.map.layers,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setView: (...params) => {
            dispatch(mapActions.setView(...params));
        },
        startMeasure: (...params) => {
            dispatch(drawingActions.startMeasure(...params));
        },
        changeMeasureComponent: (...params) => {
            dispatch(configActions.changeMeasureComponent(...params));
        },
        exportMapImage: (...params) => {
            dispatch(printActions.exportMapImage(...params));
        },
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ConfComponent));