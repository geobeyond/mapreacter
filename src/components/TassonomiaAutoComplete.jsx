import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom'
import AutoComplete from 'material-ui/AutoComplete';
import { MenuItem } from 'material-ui/Menu';
import Divider from 'material-ui/Divider';
import { newDataSourceAction } from './tassonomiaredux';

var axios = require('axios');

class TassonomiaAutoComplete extends Component {

  state = {
    dataSource: [],
    searchText: '',
  };

  handleUpdateInput = (value) => {
    console.log("TassonomiaAutoComplete.handleUpdateInput()");
    this.setState({
      searchText: value,
    });
    const url = this.props.config.tassonomiaserviceurl + value;
    console.log("GET", url);
    axios.get(url)
      .then((response) => {
        console.log("response:", JSON.stringify(response.data));
        const _datasource = [];
        this.props.config.routing.forEach(routingrecord => {
          if (response.data[routingrecord.field]) {
            if (response.data[routingrecord.field].length > 0) {
              if (_datasource.length > 0) {
                _datasource.push({ text: '', value: (<Divider />), });
              }
              _datasource.push({ text: '', value: (<MenuItem primaryText={routingrecord.label} disabled={true} />), });
              response.data[routingrecord.field].forEach(element => {
                _datasource.push({
                  text: routingrecord.routinglevel + element,
                  value: (<MenuItem primaryText={element} />),
                });
              });
            }
          }
        });
        this.props.newDataSourceAction(_datasource);
      })
      .catch((error) => {
        console.error(error);
      });
  };

  handleOnNewRequest = (chosenRequest, index) => {
    console.log("TassonomiaAutoComplete.handleOnNewRequest()", chosenRequest, index);
    this.setState({
      searchText: '',
    });
    this.props.history.push(chosenRequest.text);
  }

  render() {
    console.log("TassonomiaAutoComplete.render()");
    return (
      <AutoComplete
        style={this.props['style'] ? this.props.style : {}}
        hintText="Tassonomia ..."
        dataSource={this.props.tassonomia.dataSource}
        searchText={this.state.searchText}
        onUpdateInput={this.handleUpdateInput}
        onNewRequest={this.handleOnNewRequest}
        filter={AutoComplete.noFilter}
        openOnFocus={true}
        maxSearchResults={15}
        id={'tassonomiaautocomplete'}
        className={'tassonomiaautocomplete'}
      />
    );
  }

}

const mapStateToProps = (state) => {
  //console.log("TassonomiaAutoComplete.mapStateToProps()");
  return {
    tassonomia: state.tassonomia
  }
}

const mapDispatchToProps = {
  newDataSourceAction
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(TassonomiaAutoComplete));
