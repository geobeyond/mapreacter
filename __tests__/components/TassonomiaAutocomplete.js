import React from 'react';
import renderer from 'react-test-renderer';
import { HashRouter, Switch, Route, Link } from 'react-router-dom'
import { createStore, combineReducers, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunkMiddleware from 'redux-thunk';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { Menu, MenuItem } from 'material-ui/Menu';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TassonomiaAutoComplete from '../../src/components/TassonomiaAutoComplete';
import { tassonomiastore, newDataAction } from '../../src/components/tassonomiaredux';
import { wrap } from 'module';

configure({ adapter: new Adapter() });

const _wait = (ms) => {
    var start = new Date().getTime();
    var end = start;
    while (end < start + ms) {
        end = new Date().getTime();
    }
}

const config = {
    tassonomiaserviceurl: 'http://localhost:8080/MyRestServer/tassonomia?name=',
    routing: [
        { field: 'phylum', label: 'Phylum', routinglevel: '/' },
        { field: 'famiglia', label: 'Famiglia', routinglevel: '/*/' },
        { field: 'nome_scientifico', label: 'Specie', routinglevel: '/*/*/' },
    ],
};

test('renders correctly', () => {
    const component = renderer.create(
        <MuiThemeProvider>
            <Provider store={createStore((state = [], action) => { return state })}>
                <HashRouter>
                    <TassonomiaAutoComplete config={config} />
                </HashRouter>
            </Provider>
        </MuiThemeProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('component changes the after click', () => {

    const wrapper = mount(
        <MuiThemeProvider>
            <Provider store={createStore((state = [], action) => { return state })}>
                <HashRouter>
                    <TassonomiaAutoComplete config={config} />
                </HashRouter>
            </Provider>
        </MuiThemeProvider>
    );

    console.log('---------------------------------------------');
    console.log(wrapper.debug());
    console.log('---------------------------------------------');

    var axios = require('axios');
    var MockAdapter = require('axios-mock-adapter');
    var mock = new MockAdapter(axios);
    mock.onGet(config.tassonomiaserviceurl+'a').reply(200, {
        "phylum": [
            "Chordata",
            "Arthropoda"
        ],
        "famiglia": [
            "Emberizidae",
            "Mutillidae"
        ],
        "nome_scientifico": [
            "Asparagus acutifolius L.",
            "Crataegus monogyna Jacq.",
            "Ruscus aculeatus L.",
            "Fraxinus ornus L. subsp. ornus",
            "Tamus communis L.",
            "Bythinella schmidtii",
            "Turdus merula",
            "Sylvia atricapilla",
            "Hedera helix L. s.l.",
            "Smilax aspera L."
        ]
    });

    tassonomiastore.subscribe(() => {
        console.log('---------------------------------------------');
        config.routing.forEach(routingrecord => {
            if (tassonomiastore.getState()._data[routingrecord.field]) {
                expect(tassonomiastore.getState()._data[routingrecord.field].length).toBeGreaterThan(1);
            }
        });
        console.log('---------------------------------------------');
    });

    //wrapper.find('input').simulate('change', { target: { value: 'a' } });
    wrapper.find('#tassonomiaautocomplete').forEach((node) => {
        node.simulate('change', { target: { value: 'a' } });
    });

    _wait(60);
});
