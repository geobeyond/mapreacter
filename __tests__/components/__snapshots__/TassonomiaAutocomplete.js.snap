// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
        "muiPrepared": true,
        "position": "relative",
        "width": 256,
      }
    }
  >
    <div
      className="tassonomiaautocomplete"
      style={
        Object {
          "backgroundColor": "transparent",
          "cursor": "auto",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": 16,
          "height": 48,
          "lineHeight": "24px",
          "muiPrepared": true,
          "position": "relative",
          "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "width": 256,
        }
      }
    >
      <div
        style={
          Object {
            "bottom": 12,
            "color": "rgba(0, 0, 0, 0.3)",
            "muiPrepared": true,
            "opacity": 1,
            "position": "absolute",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          }
        }
      >
        Tassonomia ...
      </div>
      <input
        autoComplete="off"
        disabled={false}
        id="tassonomiaautocomplete"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        open={false}
        style={
          Object {
            "WebkitOpacity": 1,
            "WebkitTapHighlightColor": "rgba(0,0,0,0)",
            "appearance": "textfield",
            "backgroundColor": "rgba(0,0,0,0)",
            "border": "none",
            "color": "rgba(0, 0, 0, 0.87)",
            "cursor": "inherit",
            "font": "inherit",
            "height": "100%",
            "muiPrepared": true,
            "outline": "none",
            "padding": 0,
            "position": "relative",
            "width": "100%",
          }
        }
        type="text"
        value=""
      />
      <div>
        <hr
          aria-hidden="true"
          style={
            Object {
              "borderBottomStyle": "solid",
              "borderBottomWidth": 1,
              "borderColor": "#e0e0e0",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "bottom": 8,
              "boxSizing": "content-box",
              "margin": 0,
              "muiPrepared": true,
              "position": "absolute",
              "width": "100%",
            }
          }
        />
        <hr
          aria-hidden="true"
          style={
            Object {
              "borderBottomStyle": "solid",
              "borderBottomWidth": 2,
              "borderColor": "#00bcd4",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "bottom": 8,
              "boxSizing": "content-box",
              "margin": 0,
              "muiPrepared": true,
              "position": "absolute",
              "transform": "scaleX(0)",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
</div>
`;
